File: src/main/java/me/senseiwells/arucas/core/Lexer.java
package me.senseiwells.arucas.core;

import java.util.ArrayList;
import java.util.List;

import me.senseiwells.arucas.api.ISyntax;
import me.senseiwells.arucas.utils.Position;
import me.senseiwells.arucas.throwables.CodeError;
import me.senseiwells.arucas.tokens.Token;
import me.senseiwells.arucas.tokens.Token.Type;

public class Lexer {
	private static final LexerContext LEXER;
	
	static {
		LEXER = new LexerContext()
			// Whitespaces
			.addRule(Type.WHITESPACE, i -> i
				.addMultiline("/*", "*/")
				.addRegex("//[^\\r\\n]*")
				.addRegex("[ \t\r\n]")
			)
			
			// Arithmetics
			.addRule(Type.PLUS, i -> i.addString("+"))
			.addRule(Type.MINUS, i -> i.addString("-"))
			.addRule(Type.MULTIPLY, i -> i.addString("*"))
			.addRule(Type.DIVIDE, i -> i.addString("/"))
			.addRule(Type.POWER, i -> i.addString("^"))
			
			// Atoms
			.addRule(Type.IDENTIFIER, i -> i.addRegex("[a-zA-Z_][a-zA-Z0-9_]*"))
			.addRule(Type.BOOLEAN, i -> i.addStrings("true", "false"))
			.addRule(Type.STRING, i -> i
				.addMultiline("\"", "\\", "\"")
				.addMultiline("'", "\\", "'")
			)
			.addRule(Type.NUMBER, i -> i.addRegexes(
				"[0-9]+[.][0-9]+",
				"[0-9]+"
			))
			.addRule(Type.NULL, i -> i.addStrings("null"))

			// Comparisons - This must be defined AFTER identifiers
			.addRule(Type.EQUALS, i -> i.addString("=="))
			.addRule(Type.NOT_EQUALS, i -> i.addString("!="))
			.addRule(Type.LESS_THAN_EQUAL, i -> i.addString("<="))
			.addRule(Type.MORE_THAN_EQUAL, i -> i.addString(">="))
			.addRule(Type.LESS_THAN, i -> i.addString("<"))
			.addRule(Type.MORE_THAN, i -> i.addString(">"))
			.addRule(Type.NOT, i -> i.addStrings("!", "not"))
			.addRule(Type.AND, i -> i.addStrings("&&", "and"))
			.addRule(Type.OR, i -> i.addStrings("||", "or"))
			
			// Memory operations
			.addRule(Type.ASSIGN_OPERATOR, i -> i.addString("="))
			.addRule(Type.INCREMENT, i -> i.addString("++"))
			.addRule(Type.DECREMENT, i -> i.addString("--"))
			
			// Brackets
			.addRule(Type.LEFT_BRACKET, i -> i.addString("("))
			.addRule(Type.RIGHT_BRACKET, i -> i.addString(")"))
			.addRule(Type.LEFT_SQUARE_BRACKET, i -> i.addString("["))
			.addRule(Type.RIGHT_SQUARE_BRACKET, i -> i.addString("]"))
			.addRule(Type.LEFT_CURLY_BRACKET, i -> i.addString("{"))
			.addRule(Type.RIGHT_CURLY_BRACKET, i -> i.addString("}"))
			
			// Delimiters
			.addRule(Type.SEMICOLON, i -> i.addString(";"))
			.addRule(Type.COLON, i -> i.addString(":"))
			.addRule(Type.COMMA, i -> i.addString(","))
			
			// Keywords
			.addRule(Type.IF, i -> i.addString("if"))
			.addRule(Type.ELSE, i -> i.addString("else"))
			.addRule(Type.WHILE, i -> i.addString("while"))
			.addRule(Type.CONTINUE, i -> i.addString("continue"))
			.addRule(Type.BREAK, i -> i.addString("break"))
			.addRule(Type.RETURN, i -> i.addString("return"))
			.addRule(Type.VAR, i -> i.addString("var"))
			.addRule(Type.FUN, i -> i.addString("fun"))
			.addRule(Type.TRY, i -> i.addString("try"))
			.addRule(Type.CATCH, i -> i.addString("catch"))
			.addRule(Type.FOREACH, i -> i.addString("foreach"))
			.addRule(Type.FOR, i -> i.addString("for"))
			.addRule(Type.SWITCH, i -> i.addString("switch"))
			.addRule(Type.CASE, i -> i.addString("case"))
			.addRule(Type.DEFAULT, i -> i.addString("default"))
			.addRule(Type.CLASS, i -> i.addString("class"))
			.addRule(Type.THIS, i -> i.addString("this"))
			.addRule(Type.NEW, i -> i.addString("new"))
			.addRule(Type.STATIC, i -> i.addString("static"))
			.addRule(Type.OPERATOR, i -> i.addString("operator"))

			// Dot operator
			.addRule(Type.DOT, i -> i.addString("."))
			.addRule(Type.POINTER, i -> i.addString("->"))
		;
	}
	
	private final String text;
	private final String fileName;

	public Lexer(String text, String fileName) {
		this.text = text;
		this.fileName = fileName;
	}
	
	public List<Token> createTokens() throws CodeError {
		List<Token> tokenList = new ArrayList<>();
		int offset = 0;
		int line = 0;
		int column = 0;
		int length = this.text.length();
		String input = this.text;
		
		while (offset < length) {
			LexerContext.LexerToken lexerToken = LEXER.nextToken(input);
			
			if (lexerToken == null) {
				throw new CodeError(CodeError.ErrorType.ILLEGAL_CHAR_ERROR, "Invalid character", ISyntax.of(new Position(offset, line, column, this.fileName)));
			}
			
			if (lexerToken.length + offset > length) {
				break;
			}
			
			int old_offset = offset;
			int old_line = line;
			int old_column = column;
			
			for (int i = offset; i < offset + lexerToken.length; i++) {
				char c = this.text.charAt(i);
				
				if (c == '\n') {
					line ++;
					column = 0;
				}
				else {
					column ++;
				}
			}
			
			if (lexerToken.type != Type.WHITESPACE) {
				tokenList.add(new Token(
					lexerToken.type,
					lexerToken.content,
					new Position(old_offset, old_line, old_column, this.fileName),
					new Position(offset, line, column, this.fileName)
				));
			}
			
			input = input.substring(lexerToken.length);
			offset += lexerToken.length;
		}
	
		tokenList.add(new Token(Type.FINISH, ISyntax.of(new Position(offset, line, column, this.fileName))));
		return tokenList;
	}
}

File: src/main/kotlin/me/senseiwells/arucas/core/Lexer.kt
package me.senseiwells.arucas.core

import me.senseiwells.arucas.exceptions.compileError
import me.senseiwells.arucas.utils.LocatableTrace
import me.senseiwells.arucas.utils.StringUtils
import java.util.regex.Pattern

private val LEXER_CONTEXT = LexerContext()
    // Whitespaces
    .addRule(Type.WHITESPACE) { i ->
        i.addMultiline("/*", "*/").addRegex("//[^\\r\\n]*").addRegex("[ \t\r\n]")
    }

    // Arithmetics
    .addRule(Type.PLUS)
    .addRule(Type.MINUS)
    .addRule(Type.MULTIPLY)
    .addRule(Type.DIVIDE)
    .addRule(Type.POWER)

    // Atoms
    .addRule(Type.IDENTIFIER) { i -> i.addRegex("[a-zA-Z_][a-zA-Z0-9_]*") }
    .addRule(Type.STRING) { i -> i.addMultiline("\"", "\\", "\"").addMultiline("'", "\\", "'") }
    .addRule(Type.NUMBER) { i -> i.addRegexes("[0-9]+\\.[0-9]+", "[0-9]+", "0[xX][0-9a-fA-F]+") }
    .addRule(Type.TRUE)
    .addRule(Type.FALSE)
    .addRule(Type.NULL)

    // Comparisons - This must be defined AFTER identifiers
    .addRule(Type.EQUALS)
    .addRule(Type.NOT_EQUALS)
    .addRule(Type.LESS_THAN_EQUAL)
    .addRule(Type.MORE_THAN_EQUAL)
    .addRule(Type.LESS_THAN)
    .addRule(Type.MORE_THAN)
    .addRule(Type.NOT)
    .addRule(Type.AND)
    .addRule(Type.OR)
    .addRule(Type.XOR)
    .addRule(Type.SHIFT_LEFT)
    .addRule(Type.SHIFT_RIGHT)
    .addRule(Type.BIT_AND)
    .addRule(Type.BIT_OR)

    // Memory operations
    .addRule(Type.ASSIGN_OPERATOR)
    .addRule(Type.INCREMENT)
    .addRule(Type.DECREMENT)

    // Brackets
    .addRule(Type.LEFT_BRACKET)
    .addRule(Type.RIGHT_BRACKET)
    .addRule(Type.LEFT_SQUARE_BRACKET)
    .addRule(Type.RIGHT_SQUARE_BRACKET)
    .addRule(Type.LEFT_CURLY_BRACKET)
    .addRule(Type.RIGHT_CURLY_BRACKET)

    // Delimiters
    .addRule(Type.SEMICOLON)
    .addRule(Type.COLON)
    .addRule(Type.COMMA)

    // Keywords
    .addRule(Type.IF)
    .addRule(Type.ELSE)
    .addRule(Type.WHILE)
    .addRule(Type.CONTINUE)
    .addRule(Type.BREAK)
    .addRule(Type.RETURN)
    .addRule(Type.VAR)
    .addRule(Type.FUN)
    .addRule(Type.TRY)
    .addRule(Type.CATCH)
    .addRule(Type.FINALLY)
    .addRule(Type.FOREACH)
    .addRule(Type.FOR)
    .addRule(Type.SWITCH)
    .addRule(Type.CASE)
    .addRule(Type.DEFAULT)
    .addRule(Type.CLASS)
    .addRule(Type.ENUM)
    .addRule(Type.INTERFACE)
    .addRule(Type.THIS)
    .addRule(Type.SUPER)
    .addRule(Type.AS)
    .addRule(Type.NEW)
    .addRule(Type.STATIC)
    .addRule(Type.OPERATOR)
    .addRule(Type.THROW)
    .addRule(Type.IMPORT)
    .addRule(Type.FROM)
    .addRule(Type.LOCAL)
    .addRule(Type.ARBITRARY)
    .addRule(Type.DOT)
    .addRule(Type.POINTER)

class Lexer(private val text: String, private val fileName: String) {
    fun createTokens(): List<Token> {
        val tokenList = ArrayList<Token>()
        val length = this.text.length
        var offset = 0
        var line = 0
        var column = 0
        var input = this.text
        while (offset < length) {
            val lexerToken: LexerToken = LEXER_CONTEXT.nextToken(input) ?:
                compileError("Invalid character", LocatableTrace(this.fileName, line, column))
            if (lexerToken.length + offset > length) {
                break
            }
            val oldLine = line
            val oldColumn = column
            for (i in offset until offset + lexerToken.length) {
                val c = this.text[i]
                if (c == '\n') {
                    line++
                    column = 0
                } else {
                    column++
                }
            }
            if (lexerToken.type !== Type.WHITESPACE) {
                tokenList.add(Token(
                    lexerToken.type,
                    LocatableTrace(this.fileName, oldLine, oldColumn),
                    lexerToken.content
                ))
            }
            input = input.substring(lexerToken.length)
            offset += lexerToken.length
        }
        tokenList.add(Token(Type.EOF, LocatableTrace(this.fileName, line, column)))
        return tokenList
    }
}

private class LexerToken(val type: Type, val content: String) {
    val length: Int = this.content.length
}

/**
 * Context that stores all our lexing rules
 */
private class LexerContext {
    private val rules = ArrayList<LexerRule>()

    fun addRule(type: Type): LexerContext {
        val rule = LexerRule(type)
        rule.addString(type.toString())
        this.rules.add(rule)
        return this
    }

    fun addRule(type: Type, consumer: (LexerRule) -> Unit): LexerContext {
        val rule = LexerRule(type)
        consumer(rule)
        this.rules.add(rule)
        return this
    }

    fun nextToken(input: String): LexerToken? {
        var selectedRule: LexerRule? = null
        var longestRule = 1
        for (rule in this.rules) {
            val length = rule.getMatchLength(input)
            if (length >= longestRule) {
                longestRule = length
                selectedRule = rule
            }
        }
        return selectedRule?.let { LexerToken(selectedRule.type, input.substring(0, longestRule)) }
    }
}

/**
 * Rule that matches patterns to token types
 */
private class LexerRule(val type: Type) {
    private val matches = ArrayList<Pattern>()

    fun addString(value: String): LexerRule {
        this.matches.add(Pattern.compile(StringUtils.regexEscape(value)))
        return this
    }

    fun addRegex(regex: String): LexerRule {
        this.matches.add(Pattern.compile(regex))
        return this
    }

    fun addRegexes(vararg regexes: String): LexerRule {
        for (regex in regexes) {
            this.addRegex(regex)
        }
        return this
    }

    fun addMultiline(open: String, close: String): LexerRule {
        return this.addMultiline(open, "", close)
    }

    fun addMultiline(open: String, escape: String, close: String): LexerRule {
        val s: String = StringUtils.regexEscape(open)
        val c: String = StringUtils.regexEscape(close)
        val regex: String = if (escape.isEmpty()) "$s.*?$c" else {
            val e: String = StringUtils.regexEscape(escape)
            "$s(?:$e(?:$e|$c|(?!$c).)|(?!$e|$c).)*$c"
        }
        this.matches.add(Pattern.compile(regex, Pattern.DOTALL))
        return this
    }

    fun getMatchLength(string: String): Int {
        var length = 0
        for (pattern in this.matches) {
            val matcher = pattern.matcher(string)
            if (matcher.lookingAt()) {
                length = length.coerceAtLeast(matcher.end())
            }
        }
        return if (length < 1) -1 else length
    }
}
